homeassistant/components/backup/manager.py

```diff
@@ -774,25 +821,47 @@ async def _async_finish_backup(
                     state=CreateBackupState.IN_PROGRESS,
                 )
             )
-            agent_errors = await self._async_upload_backup(
-                backup=written_backup.backup,
-                agent_ids=agent_ids,
-                open_stream=written_backup.open_stream,
-            )
-            await written_backup.release_stream()
-            if with_automatic_settings:
-                # create backup was successful, update last_completed_automatic_backup
-                self.config.data.last_completed_automatic_backup = dt_util.now()
-                self.store.save()
-                self._update_issue_after_agent_upload(agent_errors)
+
+            try:
+                agent_errors = await self._async_upload_backup(
+                    backup=written_backup.backup,
+                    agent_ids=agent_ids,
+                    open_stream=written_backup.open_stream,
+                )
+            except BaseException:
+                self.async_on_backup_event(
+                    CreateBackupEvent(stage=None, state=CreateBackupState.FAILED)
+                )
+                raise  # manager or unexpected error
+            finally:
+                try:
+                    await written_backup.release_stream()
+                except Exception:
+                    self.async_on_backup_event(
+                        CreateBackupEvent(stage=None, state=CreateBackupState.FAILED)
+                    )
+                    raise
             self.known_backups.add(written_backup.backup, agent_errors)
+            if agent_errors:
+                self.async_on_backup_event(
+                    CreateBackupEvent(stage=None, state=CreateBackupState.FAILED)
+                )
+            else:
+                if with_automatic_settings:
+                    # create backup was successful, update last_completed_automatic_backup
+                    self.config.data.last_completed_automatic_backup = dt_util.now()
+                    self.store.save()
+
+                self.async_on_backup_event(
+                    CreateBackupEvent(stage=None, state=CreateBackupState.COMPLETED)
+                )
 
+            if with_automatic_settings:
+                self._update_issue_after_agent_upload(agent_errors)
             # delete old backups more numerous than copies
+            # try this regardless of agent errors above
             await delete_backups_exceeding_configured_count(self)
```

Comment:
Do we introduce a raise by issuing the COMPLETED event before we're really done with all operations?

Reply:
We're not idle, so I think it's ok.

Reply:
OK, so only after the idle event has been fired everything is done? I think that's fair.

Reply:
Yes until idle the manager is busy. Completed means the backup is complete. Deleting old backups is a separate operation from the backup. So everything makes sense like it is now.